gnovm/stdlibs/std/coins.gno

```diff
@@ -13,28 +14,147 @@ type Coin struct {
 	Amount int64  `json:"amount"`
 }
 
+// NewCoin returns a new coin with a denomination and amount
+func NewCoin(denom string, amount int64) Coin {
+	return Coin{
+		Denom:  denom,
+		Amount: amount,
+	}
+}
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
+	if c.IsZero() {
+		return "0" + c.Denom
+	}
+
 	return strconv.Itoa(int(c.Amount)) + c.Denom
 }
 
+// IsGTE returns true if they are the same type and the receiver is
+// an equal or greater value
 func (c Coin) IsGTE(other Coin) bool {
 	if c.Denom != other.Denom {
-		panic("invalid coin denominations: " + c.Denom)
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
 	}
+
 	return c.Amount >= other.Amount
 }
 
+// IsLT returns true if they are the same type and the receiver is
+// a smaller value
+func (c Coin) IsLT(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
+
+	return c.Amount < other.Amount
+}
+
+// IsEqual returns true if the two sets of Coins have the same value
+func (c Coin) IsEqual(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
+
+	return c.Amount == other.Amount
+}
+
+// Adds amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Add(coinB Coin) Coin {
+	res := c.AddUnsafe(coinB)
+	return res
+}
+
+func (c Coin) AddUnsafe(coinB Coin) Coin {
+	if c.Denom != coinB.Denom {
+		panic("incompatible coin denominations: " + c.Denom + ", " + coinB.Denom)
+	}
+
+	sum, ok := overflow.Add64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+
+	return Coin{c.Denom, sum}
+}
+
+// Subtracts amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Sub(coinB Coin) Coin {
+	res := c.SubUnsafe(coinB)
+	return res
+}
+
+func (c Coin) SubUnsafe(coinB Coin) Coin {
+	if c.Denom != coinB.Denom {
+		panic("incompatible coin denominations: " + c.Denom + ", " + coinB.Denom)
+	}
+	dff, ok := overflow.Sub64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+	return Coin{c.Denom, dff}
```

Comment:
Same question as with the addition, I'm not convinced we need to create the new objects every time

Reply:
Changed it: [779ff9a](https://github.com/gnolang/gno/pull/2104/commits/779ff9ad936c4e72960bbbcfe0c74628f850cd0c)